modules/core/form/js/form_protection.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * @file
+ * Show warning when a user is about to navigate away from an unsaved, dirty form.
+ */
+(function ($, Drupal) {
+
+    "use strict";
+
+    Drupal.behaviors.form_protection = {
+
+        initialFormState: {},
+
+        attach: function (context) {
+            const behavior = Drupal.behaviors.form_protection;
+            let allowClick = false;
+            let formIsDirty = false;
+
+            // Get initial form state on initial load.
+            $("form.entity-or-quick-form-protected :input", context).each(function () {
+                behavior.initialFormState[$(this).attr("id")] = $(this).serialize();
+            });
+
+            // Let all form submit buttons through.
+            $("input[type='submit'], button[type='submit']").each(function() {
+                $(this).addClass("allow-submit");
+                $(this).click(function() {
+                    allowClick = true;
+                });
+            });
+
+            // Catch all links and buttons EXCEPT for "#" links.
+            $("a, button, input[type='submit']:not(.allow-submit), button[type='submit']:not(.allow-submit)")
+                .each(function() {
+                    $(this).click(function() {
+                        formIsDirty = behavior.checkForUnsavedChanges();
+                        // If the user clicked on a # link or the form is not dirty, let click through; otherwise,
+                        // return something other than undefined to trigger onbeforeunload.
+                        if (formIsDirty && $(this).attr("href") !== "#") {
+                            return 0;
+                        }
+                    });
+                });
+
+            // Handle backbutton, exit etc.
+            window.onbeforeunload = function () {
+                if (formIsDirty && !allowClick) {
+                    allowClick = false;
+                    return (Drupal.t("You have unsaved changes."));
+                }
+            }
+        },
+
+        checkForUnsavedChanges: function () {
+            const behavior = Drupal.behaviors.form_protection;
+            let formIsDirty = false;

```

Comment:
* I think we can avoid the `formIsDirty` variable being mutable and the need to break out of the loop by using jQuery's `is` instead of `each`: https://api.jquery.com/is/ (basically akin to [Array::some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)).
* No need for the `const behavior = Drupal.behaviors.form_protection;` pattern